### ISSUE TITLE
Review current template

### ISSUE THREAD

**Issue by gentzkow** [gentzkow/template#40]
* Clone & run the template
* Look over documentation
* Let me know anything that breaks or is unclear
* Set up a time to discuss how we could make it simpler, more intuitive, or more robust

**Comment by szahedian** [comment]
@gentzkow here are my initial thoughts. I ran the repo mostly without issue. I had two small hiccups, which I describe in the first two bullets below.

Proposed edits to `README` and Wiki:
* Step 4 of setup, running `check_setup.py`, should come after step 6, installing the `gslab_make` submodule. Currently, step 4 depends on `gslab_make`, which only gets fetched in step 6.
* In `check_setup.py` line 24, the path to `__init__.py` is specified as `/lib/gslab_make/__init__.py`. However when I run this, I get a file-not-found error. When I change the path to `/lib/gslab_make/gslab_make/__init__.py`, the script successfully runs. This seems to have been changed in the other scripts that use `gslab_make`.
* It may be helpful to include a link to [Command Line Usage](https://github.com/gentzkow/template#command-line-usage) in setup step 1. An unfamiliar user may download Lyx and Stata without appending these to their path, then complete setup and wonder why their build fails. 
* It may be helpful to 
* In step 3, creating the conda environment, there could be a note open `conda_env.yaml` and change the name of the environment.
* The four required pieces of pre-setup software could include hyperlinks to download pages.
* The codeblock under "Command Line Usage" in the `README` should match `config_user_template.yaml`, so the final line should be `stata : stata-mp`. This should also be changed in the Wiki under [Configuration Files](https://github.com/gentzkow/template/wiki/Repository-Structure#configuration-files)
* I'm finding multiple links throughout the Wiki are broken.
* The Wiki [External code libraries](https://github.com/gentzkow/template/wiki/External-Dependencies#external-code-libraries) section is out of date. It references `requirements.txt` for Python, which has been superseded by conda.
* The [External file example](https://github.com/gentzkow/template/wiki/External-Dependencies#example) uses Windows directory syntax, while the rest of the Wiki defaults to Unix-like syntax

At a high level, I think the template is intuitive. One possibly vexing point may be the various ways that libraries are handled. All Stata libraries are stored in `/lib`. To the extent possible, R and Python libraries are installed through conda and pip. R libraries that cannot be installed in this way must be installed to the local non-conda environment (for me the path is `/Library/Frameworks/R.framework/Resources/library/`) using `setup_r.r`. This seems to negate some of the benefits of the isolated environment conda provides. 

I'm not sure there's an easy way of getting around this. It'd be conceptually simplest to download all libraries inaccessible through conda/pip/etc. to `/lib`. A foolproof but very heavyweight solution could be Docker containerization.

**Comment by Etang21** [comment]
@szahedian thanks so much for such a thorough runthrough and read of the template setup! Just wanted to second your bullet point (2) -- you’re totally right, that’s my bad for not catching that in https://github.com/gentzkow/template/issues/38 :) it stems from our import of `gslab_make` as a submodule earlier in the year.

The other suggestions look great to me too (and a good point on `setup_r.r` being a bit out of place), but I defer to @gentzkow on those.

**Comment by gentzkow** [comment]
@szahedian Thanks! This is excellent. 

1. All of your proposed edits to `README` and Wiki look good to me. Why don't you open a follow up issue for yourself to implement.

2. I agree with your point about `setup_r.r`. I'd be in favor of shifting the protocol to put these in `/lib` just as we do with Stata. @Etang21 @Houdanait do you foresee any problems with this?

    In a similar spirit, I wonder if we shouldn't get rid of setup_stata.do. Most of this script is designed to install extra packages into the user's local Stata installation (though in the template the list of such files is blank). We presumably prefer that people don't do this and instead add any such files to /lib/stata/.

3. I've been curious about the Docker option for a while. I could imagine this being worthwhile for big projects with heavy setup requirements. Do you think it would be worth creating a test example of how this would work so we can all get a feel? 

4. One thing I've gone back and forth on is whether the benefits of the `input.txt` structure outweigh the costs. The benefits include (i) making the dependencies among modules transparent to users, (ii) making it easy to crawl to repo and produce a DAG of the dependency flow, (iii) creating short aliases for input paths so code doesn't get cluttered w/ long paths. The cost is that it makes the whole structure less standard / a bit more clunky, and introduces a potential failure point. 

    An alternative would be to drop this and allow arbitrary calls to other modules from within code, perhaps imposing some rules so it's easy to scan / scrape these paths like requiring input paths to be declared at the top of scripts. 

    Curious if you have thoughts either way on this, or other solutions to this problem you've seen and liked.

**Comment by Houdanait** [comment]
@szahedian Thanks a lot for this! I really like your suggestions. 
On point 2 [above](https://github.com/gentzkow/template/issues/40#issuecomment-896370646), I do not anticipate any problems of this, I actually think this would make it 'official' within repositories. A few repositories had their R libraries stored in` /lib ` without appearing in the `setup_r.r`  which led to some errors when running the codes if some packages were missing. 
So I think your suggestions makes a lot of sense. (at least in my opinion!)

**Comment by Etang21** [comment]
Following up on [point 2](https://github.com/gentzkow/template/issues/40#issuecomment-896370646): Moving to use `lib/R/` instead of `setup_r.r` seems good to me! I'll add that even in a project with as much R coding as DigitalAddiction, we were able to install all R dependencies from conda, and didn't need to install any packages from [setup_r.r,](https://github.com/gentzkow/DigitalAddiction/blob/master/setup/setup_r.r). So we should have relatively few packages that need to get stored in `lib/R/`.

To check, @szahedian, for R packages that can be installed through Conda, we would still plan to install these using Conda by specifying them in `conda_env.yaml`, correct? That is my read of the comments above, but let me know if I'm mistaken.

Regarding Stata, @gentzkow did you mean replacing `download_stata_ado.do`? I believe we renamed this from `setup_stata.do`.
Replacing `download_stata_ado.do`, also sounds reasonable to me, but not quite as nice as the situation with `R`: we will have more files to carry around in `lib/stata` than in `lib/R` (see my note on DigitalAddiction above). Perhaps that is not a downside, but it feels nicer to carry only a lightweight list of packages to install, than to upload lots of library files to the repo. That said, there have been times when other RAs and I have thought our issues arose from Stata packages conflicting with already-installed packages, so keeping files in `lib/stata/` could help avoid those conflicts. Just spitballing ideas here, but is it possible for a Stata script to add specific Stata packages to a specific folder (e.g. "add this list of Stata packages to `lib/stata/`)? I'm okay with either solution here, just some thoughts.

**Comment by szahedian** [comment]
@Etang21
> To check, @szahedian, for R packages that can be installed through Conda, we would still plan to install these using Conda by specifying them in conda_env.yaml, correct?

Yes, you're right! Installing packages using conda and pip via `conda_env.yaml` is the default option. Only package unsupported by these should be downloaded in other ways.

**Comment by szahedian** [comment]
@gentzkow 

>I've been curious about the Docker option for a while. I could imagine this being worthwhile for big projects with heavy setup requirements. Do you think it would be worth creating a test example of how this would work so we can all get a feel?

Docker can automate long setups and offers perfect replicability. However, external files, returning outputs, and Stata installation may be a hassle (but [possible](https://github.com/AEADataEditor/docker-stata)). I think Docker could work well to demonstrate the replicability of relatively complex projects; maybe when a project is ready for submission, we could package it as a Docker container, ready to send off. 

I think creating an example container wouldn't be too difficult for a project that doesn't use Stata or external files. If that all sounds good to you, I'm up for taking a stab at creating such an example container.

> One thing I've gone back and forth on is whether the benefits of the input.txt structure outweigh the costs...

The modules do seem a bit clunkier with `input.txt`. I think the utility of `input.txt` depends on how modules interact in practice. If, say, a downstream module has scripts A, B, and C that all reference upstream inputs X, Y, and Z, then this added level of indirection may be helpful, in case  X, Y, and Z get renamed or reformatted into U, V, and W. But if, say, A only references X, B only Y, and C only Z, then there's little benefit to `input.txt`; in this case, updating `input.txt` is just as much work as updating paths to U, V, and W at the top of A, B, and C.

**Comment by gentzkow** [comment]
Thanks all

@Etang21: Yes, I did mean download_stata_ado.do -- sorry!

On Docker, I think the ability to work with Stata would be a must-have for it to be worth our using Docker. @szahedian Do you think it would be relatively easy to first create an example container w/out Stata then test out how clunky dealing w/ the licenses etc. for Stata would be?

On `input.txt`, I agree with that point. @Etang21 @Houdanait what are your views on this? I'd be curious in particular what your sense of the net costs vs. benefits of inputs.txt are in a big project like DigitalAddiction.

If we were to get rid of `input.txt` we would want a robust protocol for declaring and using file paths in code -- e.g., so that the (X,Y,Z) -> (U,V,W) scenario @szahedian notes is easy to handle. Do you all know of established best practices for this (esp. in R and Python)?

@szahedian One other thing I'd be curious to get your thoughts on is `/setup/config_user_template.yaml`. This has always felt like another clunky bit to me. It would be nicer if we could get rid of this and just have `config_user.yaml` at the root and have it match the template when the repo is cloned. The problem with this is we have to choose between (i) `config_user.yaml` is ignored by Git, in which case it isn't part of the repository when it's cloned or (ii) `config_user.yaml` is not ignored by Git, in which case individual users' local settings get committed / overwritten. Have you seen better solutions to this problem? I'd imagine it comes up a lot -- whenever there are local machine specific settings that users need to change within code files.

**Comment by szahedian** [comment]
@gentzkow 
1. I think creating a container first without Stata is a good approach. I can start on that.
2. I've poked around for more formal ways of doing this. I haven't turned up anything so far, but will keep looking.
3. The way we're handling `config_user.yaml` right now does seem somewhat common (see [this post](https://stackoverflow.com/questions/4743770/how-to-manage-configuration-files-when-collaborating)). Changes could be:
   1. Default config file. Have a default config file at root and the option to add an untracked config file that overrides some or all default options. Code can always fall back on the default. This would require altering how make scripts load config options. 
   2. Symlinks. Have a default config file `config_user_default.yaml` at root. As part of setup, users would have to create a symlink from `/config_user.yaml` to `config_user_default.yaml`. Alternatively, they could add their own untracked config file to root, and symlink to that. I don't this would require any changes to our make scripts.
   3. skip-worktree. Have the config file at root. As part of setup, users would run `git update-index --skip-worktree config_user.yaml`. Then git will ignore edits to the config file. See [this post](https://compiledsuccessfully.dev/git-skip-worktree/) for more info. This seems the closest to a plug-and-play solution.

**Comment by Etang21** [comment]
Neat, thanks @gentzkow for clarifying the Stata point! 

Regarding `input.txt`, I am personally of the opinion that keeping `input.txt` would be preferred to directly referencing filepaths in other modules. 

Stylistically (or psychologically?), I think we're aiming to make these modules as self-contained as possible, so that we could rip one module out of the entire repo, still have it almost run properly, and have a small number of places to fix so that it actually runs properly. In this sense, the `input.txt` file is the one central artery between this module and the rest of the modules in the repo: someone would only need to change the filepaths in `input.txt`, instead of fiddling with plumbing in every individual file.

This is partially addressed in Saam's point [here](https://github.com/gentzkow/template/issues/40#issuecomment-896499679), and my experience with Digital Addiction has been that the former is generally the case: many analysis files all reference `input/final_data.csv` in that case. I would also second (i) and (iii) in Matt's [comment here](https://github.com/gentzkow/template/issues/40#issuecomment-896370646) as quite personally convenient when developing other repos.

I could see the benefits of removing `input.txt` though, and curious for other opinions! (Note there may be alternative solutions to these two options, which I haven't thought of, that improve on both of them).

**Comment by gentzkow** [comment]
Great. Thanks all! @Houdanait let us know if your experience w/ `input.txt` matches @Etang21's above. If you concur, I'm thinking the best approach is to keep what we have but also include a note in the documentation that for very small projects with limited dependencies users may opt to remove `input.txt` and allow code to link directly.

@szahedian Thanks for all the great thoughts on the config question. Those posts you link are really helpful -- and it's great to see that what we're doing is not miles off the frontier. I really like your option (i). If I'm understanding right, we'd have a single `config.yaml` file that defines both global settings (as currently) and also defines defaults for user-specific settings (what is currently in `config_user_template.yaml`). Then a user wishing to modify a default user-specific setting does this by creating `config_user.yaml` with the overrides. 

My guess is that this wouldn't be too hard to change in `gslab_make`. @Etang21 does that match your sense?

**Comment by Houdanait** [comment]
@Etang21 @gentzkow @szahedian Thank you so so much !!

I would definetly like to keep `input.txt.`

In big projects where we try to compartmentalise as much as possible, having a file that keeps a track of all the links needed across different trees and “draws” the repository is primordial. We are then able to change the structure of the repository via this single file which proves to be extremely useful.

To be honest I cannot think of a potential con of keeping it in a big project. It’s a light file that reads easily and whose power in structuring the repository is great.

**Comment by gentzkow** [comment]
Great. Thanks @Etang21.

@szahedian Would you be up for opening and issue for yourself to update that `gslab_make` function? It will be useful in any case for you to get familiar w/ that code.

**Comment by szahedian** [comment]
Yup @gentzkow I'm happy to work on that! Just created the [issue](https://github.com/gslab-econ/gslab_make/issues/41).

**Comment by szahedian** [comment]
@gentzkow could you give me contributor permission to [gslab_make](https://github.com/gslab-econ/gslab_make)? Right now I'm unable to create branches.

**Comment by gentzkow** [comment]
Sure. Should be good now.

>

**Comment by gentzkow** [comment]
@szahedian Checking in on status of this issue. If you feel like the goals of this issue have been completed (seems to me you've definitely reviewed the template!) you can go ahead and close w/ a final comment per [here](https://github.com/gslab-econ/ra-manual/wiki/Issues#closing-an-issue).

**Comment by szahedian** [comment]
**Summary:** Several members of the team discussed improvements to the template. The discussion spawned several new issues: #43 and gslab-econ/gslab_make#41 followed directly from this discussion; #44 and #45 followed indirectly.
